### PR TITLE
Feature/pagination

### PR DESCRIPTION
**Pull request by willarion**
made pagination

### CODE REVIEW

File: src/api/api.ts [hunk 1]
```diff
@@ -3,26 +3,35 @@ import { GoogleMessage } from '../models/GoogleMessage';
...
+  messages: Array<string>;
+  pageToken: string;
+}
+
 export const getMessages = async (
   authToken: string,
   filter = '',
-  // pageToken?: string,
-): Promise<Array<string>> => {
+  nextPageToken = '0',
```
Comment (bondarchukse) [inline src/api/api.ts]: it should be null!

File: src/api/api.ts [hunk 2]
```diff
@@ -40,17 +49,26 @@ export const getMessageContent = async (
...
   // todo
   filter = '',
-  // pageToken?: string,
-): Promise<Array<UserMessage>> => {
-  const ids = await getMessages(authToken, filter);
+  nextPageToken?: string,
+): Promise<UserMessagesInfo> => {
+  const messagesInfo = await getMessages(authToken, filter, nextPageToken);
+  const ids = messagesInfo.messages;
   const messages = ids.map((id) => getMessageContent(id, authToken));
```
Comment (bondarchukse) [inline src/api/api.ts]: lodash `map`!

File: src/api/api.ts [hunk 3]
```diff
@@ -40,17 +49,26 @@ export const getMessageContent = async (
...
+
   const result = await Promise.allSettled(messages);
-  return result
+  const finalResult = result
```
Comment (bondarchukse) [inline src/api/api.ts]: `finalResult` - think about name!

File: src/api/api.ts [hunk 4]
```diff
@@ -40,17 +49,26 @@ export const getMessageContent = async (
...
 export const getMessagesList = async (
   authToken: string,
   // todo
   filter = '',
-  // pageToken?: string,
-): Promise<Array<UserMessage>> => {
-  const ids = await getMessages(authToken, filter);
+  nextPageToken?: string,
+): Promise<UserMessagesInfo> => {
+  const messagesInfo = await getMessages(authToken, filter, nextPageToken);
```
Comment (bondarchukse) [inline src/api/api.ts]: `const { messages, pageToken } = await getMessages(authToken, filter, nextPageToken);`

File: src/components/PostsContainer/PostsContainer.tsx [hunk 5]
```diff
@@ -7,19 +7,39 @@ import { Search } from '../Search/Search';
...
+  const memorizedFilter = React.useMemo(
+    () => makeGoogleSearchQuery(search),
+    [search],
+  );
+
+  const { messages, pageToken, getNextMessagesList, getPreviousMessagesList } =
+    useMessages(memorizedFilter);
 
-  const messages = useMessages(makeGoogleSearchQuery(search));
+  const goForward = () => {
```
Comment (bondarchukse) [inline src/components/PostsContainer/PostsContainer.tsx:27]: `useCallback`

File: src/components/PostsContainer/PostsContainer.tsx [hunk 6]
```diff
@@ -7,19 +7,39 @@ import { Search } from '../Search/Search';
...
+    getNextMessagesList(memorizedFilter);
+  };
+  const goBack = () => {
+    getPreviousMessagesList(memorizedFilter);
+  };
+  const isBackButtonActive = !(pageToken.length > 1);
```
Comment (bondarchukse) [inline src/components/PostsContainer/PostsContainer.tsx]: if `pageToken` is string then 
```
 <NavigationButtons
          onBackButton={!pageToken}
          onBack={goBack}
          onForward={goForward}
        />
```

File: src/components/PostsContainer/PostsContainer.tsx [hunk 7]
```diff
@@ -7,19 +7,39 @@ import { Search } from '../Search/Search';
...
 
   return (
     <section
       className={classNames('postsContainer', styles.postsContainer_alone)}
     >
-      <Search handleInput={handleInput} input={input} />
+      <div className={classNames(styles.tools)}>
+        <Search handleInput={handleInput} input={input} />
+        <NavigationButtons
+          onBackButton={isBackButtonActive}
```
Comment (bondarchukse) [inline src/components/PostsContainer/PostsContainer.tsx]: we are using prefix `on` only for actions!

File: src/hooks/useMessages.ts [hunk 8]
```diff
@@ -2,18 +2,55 @@ import React from 'react';
...
 
   const [messages, setMessages] = React.useState<Array<UserMessage>>([]);
+
+  const { pageToken, saveOnePageToken, saveLessPageTokens, saveMorePageToken } =
+    useNextPageToken();
+
   React.useEffect(() => {
     getMessagesList(token, filter)
-      .then((res) => setMessages(res))
+      .then((res) => {
```
Comment (bondarchukse) [inline src/hooks/useMessages.ts]: `.then(({ finalResult, pageToken }) => {`

File: src/hooks/useMessages.ts [hunk 9]
```diff
@@ -2,18 +2,55 @@ import React from 'react';
...
+        setMessages(res.finalResult);
+        saveOnePageToken(res.pageToken);
+      })
       .catch(() => setMessages([]));
   }, [filter]);
 
-  return messages;
+  const getNextMessagesList = (oldFilter: string) => {
```
Comment (bondarchukse) [inline src/hooks/useMessages.ts]: `useCallback`!

File: src/hooks/useNextPageToken.ts [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+export const useNextPageToken = (): {
+  pageToken: Array<string>;
+  saveOnePageToken: (apiPageToken: string) => void;
+  saveMorePageToken: (apiPageToken: string) => void;
+  saveLessPageTokens: (lessTokens: Array<string>) => void;
+} => {
+  const [pageToken, setPageToken] = React.useState<Array<string>>(['0']);
+
+  const saveOnePageToken = (apiPageToken: string) => {
```
Comment (bondarchukse) [inline src/hooks/useNextPageToken.ts]: useCallback!

File: src/hooks/useNextPageToken.ts [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+    setPageToken([apiPageToken]);
+  };
+
+  const saveMorePageToken = (apiPageToken: string) => {
+    setPageToken([...pageToken, apiPageToken]);
+  };
+
+  const saveLessPageTokens = (lessTokens: Array<string>) => {
```
Comment (bondarchukse) [inline src/hooks/useNextPageToken.ts]: function do nothing!
```
return { pageToken, saveOnePageToken, saveMorePageToken, saveLessPageTokens: setPageToken };
````

File: src/components/PostsContainer/PostsContainer.tsx (hunk 6 above)
Comment (willarion) [reply src/components/PostsContainer/PostsContainer.tsx]: pageToken is an array of strings, I'll rename it to pageTokensList

File: src/components/PostsContainer/PostsContainer.tsx (hunk 5 above)
Comment (willarion) [reply src/components/PostsContainer/PostsContainer.tsx:27]: I've done it like this after our yesterday's standup:
`  const memorizedFilter = React.useMemo(
    () => makeGoogleSearchQuery(search),
    [search],
  );
`

is it okay?

**Pull request merged**